### PR TITLE
Rewrite in ES6

### PR DESCRIPTION
**Pull request by dougmoscrop**
Sooo the jscs rules didn't pass on the existing code, I'm really confused about how that happened, but anyway, I "fixed" most of them (except the one rule I disabled, which I could not make happy).

I fixed a bug in the original impl (it used `.uri` instead of `.url`) and ensured AWS config was loaded.

### CODE REVIEW

File: lib/aws-sigv4.js [hunk 1]
```diff
@@ -11,118 +11,123 @@ var aws = require('aws-sdk'),
...
-                }
+    constructor(scriptConfig, eventEmitter) {
+        this.validateScriptConfig(scriptConfig);
+        this.serviceName = scriptConfig.plugins[constants.PLUGIN_NAME][constants.PLUGIN_PARAM_SERVICE_NAME];
 
-                if (requestParams.body) {
-                    req.body = requestParams.body;
-                } else if (requestParams.json) {
-                    req.body = JSON.stringify(requestParams.json);
+        this.config = new Promise((resolve, reject) => {
```
Comment (dougmoscrop) [inline lib/aws-sigv4.js]: this is a lazy-load but only load once rather than on each request

File: lib/aws-sigv4.js [hunk 2]
```diff
@@ -11,118 +11,123 @@ var aws = require('aws-sdk'),
...
+        }
+        return false;
+    }
+
+    addAmazonSignatureV4(requestParams, context, callback) {
+        Promise.resolve(this.config)
+            .then(config => {
+                const credentials = config.credentials,
+                    region = config.region,
+                    targetUrl = url.parse(requestParams.url),
```
Comment (dougmoscrop) [inline lib/aws-sigv4.js]: this was `.uri` and failed with me on latest artillery

**Comment by erikerikson** [comment]
Sorry for the delay and thank you for this.  Taking a look.

**Comment by erikerikson** [comment]
And explanation of that rule is here: http://jscs.info/rule/requireMultipleVarDecl.html

Basically, it declares the declaration of variables like so:
```
var a,
    b,
    c;
```
And at the top of the scope.

We've moved on, generally speaking from these rules, using eslint and a flavor thereof...  so apologies for the broken state of master and legacy.  Not sure how that happened but thanks for working past it.

There's a lot of refactor here.

**Comment by erikerikson** [comment]
LTS for 4 doesn't end until April 2018.

I am thereby of the opinion that an es6 refactor should be held of until then.  @gwsii as the other maintainer here, do you have an opinion to add?

I'm happy to fast track a fix of `uri` => `url`.  That change was introduced here: https://github.com/shoreditch-ops/artillery-core/commit/48dfbe91d67642fa1c93989acd95af018e6f42c4#diff-08c61297a57965f46e6bf18087f2066b

**Comment by erikerikson** [comment]
@dougmoscrop between pulls https://github.com/Nordstrom/artillery-plugin-aws-sigv4/pull/10 and https://github.com/Nordstrom/artillery-plugin-aws-sigv4/pull/11 I believe the issues here should be resolved in a compatibiltiy maximizing fashion.  Have I missed something that was changed here?  Due to the refactor, I had trouble ensuring every change was reflected.  Thanks for calling this to attention and offering a solution!

**Comment by erikerikson** [comment]
Oh, right, thank you for calling that out.  We've previously depended upon environment credentials, particularly since we usually use this inside Lambdas but this will increase the usability.  Thanks!

**Comment by erikerikson** [comment]
FYI @dougmoscrop we have published these changes as v0.0.3.

Thank you for bringing attention to this!

**Comment by dougmoscrop** [comment]
Thanks!

**Comment by erikerikson** [comment]
Of course.  Sorry for the original problem as well as the delay

**Pull request closed without merging**